receiver/factorytemplate/factory_template.go

```diff
@@ -0,0 +1,186 @@
+		return nil, ErrNilNewDefaultCfg
+	}
+	if newReceiver == nil {
+		return nil, ErrNilNewReceiver
+	}
+
+	return &traceReceiverFactory{
+		factory: factory{
+			receiverType:  receiverType,
+			newDefaultCfg: newDefaulfCfg,
+		},
+		newReceiver: newReceiver,
+	}, nil
+}
+
+// NewMetricsReceiverFactory creates a factory for the given receiver "type" that
+// will have as the default configuration the object returned by newDefaultCfg and
+// that can be created using newReceiver from the configuration type returned by
+// newDefaultCfg.
+//
+// The object returned by newDefaultCfg should have its fields properly decorated using
+// the mapstructure attribute.
+//
+// The first parameter passed to newReceiver is going to be from the same type returned by
+// newDefaultCfg. The receiver implementer is going to be able to cast it to the appropriate
+// type.
+func NewMetricsReceiverFactory(
+	receiverType string,
+	newDefaulfCfg func() interface{},
+	newReceiver func(interface{}) (receiver.MetricsReceiver, error),
+) (receiver.MetricsReceiverFactory, error) {
+	if receiverType == "" {
+		return nil, ErrEmptyReciverType
+	}
+	if newDefaulfCfg == nil {
+		return nil, ErrNilNewDefaultCfg
+	}
+	if newReceiver == nil {
+		return nil, ErrNilNewReceiver
+	}
+
+	return &metricsReceiverFactory{
+		factory: factory{
+			receiverType:  receiverType,
+			newDefaultCfg: newDefaulfCfg,
+		},
+		newReceiver: newReceiver,
+	}, nil
+}
+
+// Type gets the type of the receiver created by this factory.
+func (f *factory) Type() string {
+	return f.receiverType
+}
+
+// NewFromViper takes a viper.Viper configuration and creates a new TraceReceiver.
+// Returning also the configuration used to create it.
+func (trf *traceReceiverFactory) NewFromViper(v *viper.Viper) (receiver.TraceReceiver, interface{}, error) {
+	cfg, err := trf.configFromViper(v)
+	if err != nil {
+		return nil, nil, err
+	}
+	r, err := trf.newReceiver(cfg)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return r, cfg, nil
+}
+
+// NewFromViper takes a viper.Viper configuration and creates a new TraceReceiver.
+// Returning also the configuration used to create it.
+func (mrf *metricsReceiverFactory) NewFromViper(v *viper.Viper) (receiver.MetricsReceiver, interface{}, error) {
+	cfg, err := mrf.configFromViper(v)
+	if err != nil {
+		return nil, nil, err
+	}
+	r, err := mrf.newReceiver(cfg)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return r, cfg, nil
+}
+
+// AddDefaultConfig returns the default configuration for the receiver
+// created by this factory.
+func (f *factory) AddDefaultConfig(v *viper.Viper) {
+	v.SetDefault("receivers."+f.Type(), f.newDefaultCfg())
+}
+
+// configFromViper takes a viper.Viper, generates a default config and returns the
+// resulting configuration.
+func (f *factory) configFromViper(v *viper.Viper) (cfg interface{}, err error) {
+	if v == nil {
+		return nil, ErrNilViper
+	}
+
+	cfg = f.newDefaultCfg()
+	err = v.UnmarshalKey("receivers."+f.Type(), cfg)
```

Comment:
Not sure how Viper works but was expecting that we pass only the subtree of the config to the factory, and factories do not know about the the whole structure of the tree, I saw `Sub` function that does that, similar for retrieving the default config was expecting that we return the subtree of the config that is specific to this factory.

The core config module will build the entire tree, and parse the entire tree in subtrees. This way we isolate where the structure of the tree lives, currently every receiver/exporter/processor needs to know about the whole tree, which will make impossible to change in the future.

Reply:
Yes, this is not correct and not needed (assuming caller passing the proper viper sub).

Reply:
Agreed. We should only pass the subtrees, I believe `ExportersFromViperConfig` does this sort of thing in `internal/config/config.go` as a potential example of the use of the `Sub` that @bogdandrutu mentioned.

Reply:
Also I think the `DefaultConfig() *viper.Viper` is better because we just return the subtree specific for this factory.

Reply:
I didn't see a direct way to merge the various viper.Viper with different configurations so if we return a new viper for each DefaultConfig the onus is on us to properly merge that information.

Reply:
A related issue is that our discussed usage of array of items doesn't go straight with v.Sub as we used in other places (or at least with the APIs that I've seen from Viper)

Reply:
Maybe Viper is not the right tool then. I think from a design point of view the receiver/exporter should not know about the whole structure of our config tree, otherwise will be in a huge problem later because we encode this structure in all the components.

Reply:
Only the factories needs to deal with viper, receivers/exporters don't need to have anything to deal with viper. That said it is true that the factories are having a side effect on the viper instance. For reading the config I like viper but it is really not made to output a config file like we want to do for default, one alternative is for us to marshal the configs directly.